server/db/db.js

```diff
@@ -34,488 +34,654 @@ function createHubFromRow(r) {
 }
 
 module.exports.beginTransaction = async function() {
-    let client = await pool.connect()
-    await client.query("BEGIN")
-    
-    // this fixes a race condition when two button press messages are received in quick succession
-    // this means that only one transaction executes at a time, which is not good for performance
-    // we should revisit this when / if db performance becomes a concern
-    await client.query("LOCK TABLE sessions, registry, installations, migrations")
-    
+    let client = null
+    try {
+        client = await pool.connect()
+        await client.query("BEGIN")
+        
+        // this fixes a race condition when two button press messages are received in quick succession
+        // this means that only one transaction executes at a time, which is not good for performance
+        // we should revisit this when / if db performance becomes a concern
+        await client.query("LOCK TABLE sessions, registry, installations, migrations")
+        
+        return client
+    }
+    catch(e) {
+        helpers.log(`Error running the beginTransaction query: ${e}`);
+    }
+
     return client
 }
 
 module.exports.commitTransaction = async function(client) {
-    await client.query("COMMIT")
-    client.release()
-}
-
-module.exports.getUnrespondedSessionWithButtonId = async function(buttonId, client) {
-    let transactionMode = (typeof client !== 'undefined')
-    if(!transactionMode) {
-        client = await pool.connect()
+    try {
+        await client.query("COMMIT")
+        client.release()
+    }
+    catch(e) {
+        helpers.log(`Error running the commitTransaction query: ${e}`);
     }
+}
 
-    const query = "SELECT * FROM sessions WHERE button_id = $1 AND state != $2 AND state != $3 AND state != $4"
-    const values = [buttonId, ALERT_STATE.WAITING_FOR_CATEGORY, ALERT_STATE.WAITING_FOR_DETAILS, ALERT_STATE.COMPLETED]
-    const { rows } = await client.query(query, values)
-   
-    if(!transactionMode) {
+module.exports.rollbackTransaction = async function(client) {
+    try {
+        await client.query("ROLLBACK")
         client.release()
     }
+    catch(e) {
+        helpers.log(`Error running the rollbackTransaction query: ${e}`);
+    }
+}
+
+module.exports.getUnrespondedSessionWithButtonId = async function(buttonId, client) {
+    try {
+        let transactionMode = (typeof client !== 'undefined')
+        if(!transactionMode) {
+            client = await pool.connect()
+        }
 
-    if(rows.length > 0) {        
-        return createSessionFromRow(rows[0])
+        const query = "SELECT * FROM sessions WHERE button_id = $1 AND state != $2 AND state != $3 AND state != $4"
+        const values = [buttonId, ALERT_STATE.WAITING_FOR_CATEGORY, ALERT_STATE.WAITING_FOR_DETAILS, ALERT_STATE.COMPLETED]
+        const { rows } = await client.query(query, values)
+    
+        if(!transactionMode) {
+            client.release()
```

Comment:
if we're not in transaction mode, but an error gets thrown during the SELECT query above, the code path will skip this `client.release()` call. this might lead to our db client pool running out of clients to use.

Reply:
Done.